hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java

```diff
@@ -173,71 +170,59 @@ public static synchronized void executeJar(@Nonnull Supplier<HazelcastInstance>
             }
 
         } finally {
-            // HazelcastInstance should be closed if called by client side
-            // HazelcastInstance should not be closed if called by member side
+            // HazelcastInstance should be closed only if called by the client side
             if (!calledByMember) {
                 HazelcastInstance remembered = HazelcastBootstrap.supplier.remembered();
                 if (remembered != null) {
                     try {
                         remembered.shutdown();
-                    } catch (Throwable t) {
-                        System.err.println("Shutdown failed with:");
-                        t.printStackTrace();
+                    } catch (Exception exception) {
+                        LOGGER.severe("Shutdown failed with:", exception);
                     }
                 }
                 resetSupplier();
             }
         }
     }
 
-    static String getMainClass(String mainClass, String jarPath, boolean calledByMember)
+    static String findMainClassNameForJar(String mainClass, String jarPath, boolean calledByMember)
             throws IOException {
-        MainClassFinder mainClassFinder = new MainClassFinder();
-        mainClassFinder.findMainClass(mainClass, jarPath, calledByMember);
+        MainClassNameFinder mainClassNameFinder = new MainClassNameFinder();
+        mainClassNameFinder.findMainClass(mainClass, jarPath, calledByMember);
 
         // Check if there is an error
-        String errorMessage = mainClassFinder.getErrorMessage();
+        String errorMessage = mainClassNameFinder.getErrorMessage();
         if (!StringUtil.isNullOrEmpty(errorMessage)) {
+            // Client side immediately exits, server side throws JetException
             error(errorMessage, calledByMember);
         }
-
-        // Check if mainClassName found
-        String mainClassName = mainClassFinder.getMainClassName();
-        if (StringUtil.isNullOrEmpty(mainClassName)) {
-            error(errorMessage, calledByMember);
-        }
-        return mainClassName;
-    }
-
-    // Returns true if the main method is both public and static
-    static boolean isPublicAndStatic(Method mainMethod) {
-        int modifiers = mainMethod.getModifiers();
-        return isPublic(modifiers) && isStatic(modifiers);
+        return mainClassNameFinder.getMainClassName();
     }
 
-    static Method getMainMethod(Class<?> clazz, boolean calledByMember) throws NoSuchMethodException {
-        // If main method does not exist, throws NoSuchMethodException
-        Method main = clazz.getDeclaredMethod("main", String[].class);
+    static Method findMainMethodForJar(ClassLoader classLoader, String mainClassName, boolean calledByMember) {
+        MainMethodFinder mainMethodFinder = new MainMethodFinder();
```

Comment:
I'd refactor it like below, WDYT?

```java
Subject: [PATCH] Changed window size just to make it different from input object count
---
Index: hazelcast/src/test/java/com/hazelcast/instance/impl/MainMethodFinderTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hazelcast/src/test/java/com/hazelcast/instance/impl/MainMethodFinderTest.java b/hazelcast/src/test/java/com/hazelcast/instance/impl/MainMethodFinderTest.java
--- a/hazelcast/src/test/java/com/hazelcast/instance/impl/MainMethodFinderTest.java	(revision 823bf92f4bbd2813fa83575906c8e0909b37669e)
+++ b/hazelcast/src/test/java/com/hazelcast/instance/impl/MainMethodFinderTest.java	(date 1677163571741)
@@ -16,6 +16,7 @@
 
 package com.hazelcast.instance.impl;
 
+import com.hazelcast.instance.impl.MainMethodFinder.Result;
 import org.junit.Test;
 
 import java.lang.reflect.Method;
@@ -31,29 +32,16 @@
     }
 
     @Test
-    public void testPublicAndStaticForMain() throws NoSuchMethodException {
-        Method method = MainMethodFinderTest.class.getDeclaredMethod("main", String[].class);
-
-        MainMethodFinder mainMethodFinder = new MainMethodFinder();
-        mainMethodFinder.setMainMethod(method);
-        boolean publicAndStatic = mainMethodFinder.isPublicAndStatic();
-        assertTrue(publicAndStatic);
-    }
-
-    @Test
-    public void testPublicAndStaticForSelf() throws NoSuchMethodException {
-        Method method = MainMethodFinderTest.class.getDeclaredMethod("testPublicAndStaticForSelf");
-
-        MainMethodFinder mainMethodFinder = new MainMethodFinder();
-        mainMethodFinder.setMainMethod(method);
-        boolean publicAndStatic = mainMethodFinder.isPublicAndStatic();
-        assertFalse(publicAndStatic);
+    public void testPublicAndStaticForMain() {
+        Result result = MainMethodFinder.findMainMethod(this.getClass().getClassLoader(),
+                this.getClass().getName());
+        assertFalse(result.hasError());
     }
 
     @Test
     public void testGetMainMethod() {
-        MainMethodFinder mainMethodFinder = new MainMethodFinder();
-        mainMethodFinder.getMainMethodOfClass(MainMethodFinderTest.class);
-        assertNotNull(mainMethodFinder.getMainMethod());
+        Result result = MainMethodFinder.findMainMethod(this.getClass().getClassLoader(),
+                this.getClass().getName());
+        assertNotNull(result.mainMethod);
     }
 }
Index: hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java b/hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java
--- a/hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java	(revision 823bf92f4bbd2813fa83575906c8e0909b37669e)
+++ b/hazelcast/src/main/java/com/hazelcast/instance/impl/HazelcastBootstrap.java	(date 1677163571738)
@@ -37,6 +37,7 @@
 import com.hazelcast.crdt.pncounter.PNCounter;
 import com.hazelcast.durableexecutor.DurableExecutorService;
 import com.hazelcast.flakeidgen.FlakeIdGenerator;
+import com.hazelcast.instance.impl.MainMethodFinder.Result;
 import com.hazelcast.internal.util.StringUtil;
 import com.hazelcast.jet.JetCacheManager;
 import com.hazelcast.jet.JetException;
@@ -200,16 +201,14 @@
     }
 
     static Method findMainMethodForJar(ClassLoader classLoader, String mainClassName, boolean calledByMember) {
-        MainMethodFinder mainMethodFinder = new MainMethodFinder();
-        mainMethodFinder.findMainMethod(classLoader, mainClassName);
+        Result method = MainMethodFinder.findMainMethod(classLoader, mainClassName);
 
         // Check if there is an error
-        String errorMessage = mainMethodFinder.getErrorMessage();
-        if (!StringUtil.isNullOrEmpty(errorMessage)) {
+        if (method.hasError()) {
             // Client side immediately exits, server side throws JetException
-            error(errorMessage, calledByMember);
+            error(method.errorMessage, calledByMember);
         }
-        return mainMethodFinder.getMainMethod();
+        return method.mainMethod;
     }
 
     private static void resetJetParametersForMember(String jar, String snapshotName, String jobName) {
Index: hazelcast/src/main/java/com/hazelcast/instance/impl/MainMethodFinder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/hazelcast/src/main/java/com/hazelcast/instance/impl/MainMethodFinder.java b/hazelcast/src/main/java/com/hazelcast/instance/impl/MainMethodFinder.java
--- a/hazelcast/src/main/java/com/hazelcast/instance/impl/MainMethodFinder.java	(revision 823bf92f4bbd2813fa83575906c8e0909b37669e)
+++ b/hazelcast/src/main/java/com/hazelcast/instance/impl/MainMethodFinder.java	(date 1677163627030)
@@ -16,6 +16,8 @@
 
 package com.hazelcast.instance.impl;
 
+import com.hazelcast.internal.util.StringUtil;
+
 import java.lang.reflect.Method;
 
 import static java.lang.reflect.Modifier.isPublic;
@@ -23,48 +25,43 @@
 
 class MainMethodFinder {
 
-    private String errorMessage;
-
-    private Method mainMethod;
+    static class Result {
+        final String errorMessage;
 
-    public String getErrorMessage() {
-        return errorMessage;
-    }
+        final Method mainMethod;
 
-    public Method getMainMethod() {
-        return mainMethod;
-    }
-
-    // Public for testing purpose
-    public void setMainMethod(Method mainMethod) {
-        this.mainMethod = mainMethod;
-    }
+        public Result(String errorMessage, Method mainMethod) {
+            this.errorMessage = errorMessage;
+            this.mainMethod = mainMethod;
+        }
 
-    public void findMainMethod(ClassLoader classLoader, String mainClassName) {
+        public boolean hasError() {
+            return !StringUtil.isNullOrEmpty(errorMessage);
+        }
+    }
+
+    public static Result findMainMethod(ClassLoader classLoader, String mainClassName) {
         try {
             Class<?> clazz = classLoader.loadClass(mainClassName);
-            getMainMethodOfClass(clazz);
+            return getMainMethodOfClass(clazz);
         } catch (ClassNotFoundException e) {
-            errorMessage = "Cannot find or load main class: " + mainClassName;
+            return new Result("Cannot find or load main class: " + mainClassName, null);
         }
     }
 
-    void getMainMethodOfClass(Class<?> clazz) {
+    private static Result getMainMethodOfClass(Class<?> clazz) {
         try {
-            mainMethod = clazz.getDeclaredMethod("main", String[].class);
+            Method mainMethod = clazz.getDeclaredMethod("main", String[].class);
 
-            if (!isPublicAndStatic()) {
-                errorMessage = "Class " + clazz.getName()
-                               + " has a main(String[] args) method which is not public static";
-            }
-        } catch (NoSuchMethodException exception) {
-            errorMessage = "Class " + clazz.getName()
-                           + " does not have a main(String[] args) method";
-        }
-    }
-
-    boolean isPublicAndStatic() {
-        int modifiers = mainMethod.getModifiers();
-        return isPublic(modifiers) && isStatic(modifiers);
+            int modifiers = mainMethod.getModifiers();
+            if (!(isPublic(modifiers) && isStatic(modifiers))) {
+                return new Result("Class " + clazz.getName()
+                        + " has a main(String[] args) method which is not public static", null);
+            }
+            return new Result(null, mainMethod);
+        } catch (NoSuchMethodException exception) {
+            return new Result("Class " + clazz.getName()
+                    + " does not have a main(String[] args) method", null);
+        }
     }
 }

```

Reply:
Ok. Good idea. I have added a nested Result object to MainMethodFinder